torchaudio/functional/filtering.py

```diff
@@ -642,6 +642,44 @@ def equalizer_biquad(
     return biquad(waveform, b0, b1, b2, a0, a1, a2)
 
 
+def filtfilt(
+    waveform: Tensor,
+    a_coeffs: Tensor,
+    b_coeffs: Tensor,
+    clamp: bool = True,
+    batching: bool = True,
+) -> Tensor:
+    r"""Perform an IIR filter forward and backward to a waveform.
+
+    Inspired by https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.filtfilt.html
+
+    Args:
+        waveform (Tensor): audio waveform of dimension of ``(..., time)``.  Must be normalized to -1 to 1.
+        a_coeffs (Tensor): denominator coefficients of difference equation of dimension of either
+                                1D with shape ``(num_order + 1)`` or 2D with shape ``(num_filters, num_order + 1)``.
+                                Lower delays coefficients are first, e.g. ``[a0, a1, a2, ...]``.
+                                Must be same size as b_coeffs (pad with 0's as necessary).
+        b_coeffs (Tensor): numerator coefficients of difference equation of dimension of either
+                                1D with shape ``(num_order + 1)`` or 2D with shape ``(num_filters, num_order + 1)``.
+                                Lower delays coefficients are first, e.g. ``[b0, b1, b2, ...]``.
+                                Must be same size as a_coeffs (pad with 0's as necessary).
+        clamp (bool, optional): If ``True``, clamp the output signal to be in the range [-1, 1] (Default: ``True``)
+        batching (bool, optional): Effective only when coefficients are 2D. If ``True``, then waveform should be at
+                                    least 2D, and the size of second axis from last should equals to ``num_filters``.
+                                    The output can be expressed as ``output[..., i, :] = lfilter(waveform[..., i, :],
+                                    a_coeffs[i], b_coeffs[i], clamp=clamp, batching=False)``. (Default: ``True``)
+
+    Returns:
+        Tensor: Waveform with dimension of either ``(..., num_filters, time)`` if ``a_coeffs`` and ``b_coeffs``
+                are 2D Tensors, or ``(..., time)`` otherwise.
+    """
+    forward_filtered = lfilter(waveform, a_coeffs, b_coeffs, batching=batching)
+    backward_filtered = lfilter(
+        forward_filtered.flip(-1), a_coeffs, b_coeffs, clamp=clamp, batching=batching
+    ).flip(-1)
```

Comment:
After the first pass of lfilter, the waveform shape might have changed because of using filterbanks, if this case happened, the second pass should be done in batching mode:
```suggestion
    forward_filtered = lfilter(waveform, a_coeffs, b_coeffs, clamp=False, batching=batching)
    if forward_filtered.shape != waveform.shape:
        batching = True
    backward_filtered = lfilter(
        forward_filtered.flip(-1), a_coeffs, b_coeffs, clamp=clamp, batching=batching
    ).flip(-1)
```

Reply:
I'm curious about the conditional logic here. If we set ``batching`` to ``False`` in filtfilt, will the shape of ``forward_filtered`` still be different to ``waveform``?

Reply:
It can still be different if the coefficients are 2D Tensor, thus operate in filterbanks mode, result in one extra dimension.

Reply:
I checked the lfilter impl. If ``waveform`` is 1D, ``a_coeffs`` is 2D, ``batching`` has to be ``False``, otherwise it will throw an error like ``assert waveform.ndim > 1``. And the output is 2D.
If ``batching`` is ``True``, then ``waveform`` has to be 2D, so is ``a_coeffs``, and the output is also 2D.
In the first case the output shape can be different from waveform. That makes sense to me.